File: day01/solution.py
#!/usr/bin/env python3
#
# 2025 Advent of Code, Day 1
# 2025-12-02 Stephen Houser <stephenhouser@gmail.com>
# 
import time     # for performance_counter()
import sys      # for sys.argv

def read_input(file: str):
    """Read input data file, return list of turns"""

    data = []

    with open(file, "r", encoding="utf-8") as inf:
        for line in inf.readlines():
            direction = -1 if line[0] == "L" else 1
            data.append(direction * int(line[1:]))

    return data

def turn_dial(dial, steps):
    """Turns the dial from start by steps and returns the new position"""

    # reverse the world on left turns
    dial = (100 - dial if steps < 0 else dial) % 100

    # turn the dial, count zero crossings and set the new dial position
    next = dial + abs(steps)
    zeros = next // 100
    dial = next % 100

    # reverse the world back on left turns
    dial = (100 - dial if steps < 0 else dial) % 100
    return (dial, zeros)

def part1(turns, start):
    """Returns the number of times the dial is left on 0 after turning."""

    dial = start
    zeros = 0

    for steps in turns:
        (dial, _) = turn_dial(dial, steps)

        if dial == 0:
            zeros += 1

    return zeros

def part2(turns, start):
    """Returns the number of times the dial crosses 0 while turning."""

    dial = start
    zero_crossings = 0

    for steps in turns:
        (dial, zeros) = turn_dial(dial, steps)
        zero_crossings += zeros

    return zero_crossings


def main():
    if len(sys.argv) == 1:
        print("No input file given")
        exit(1)

    start_time = time.perf_counter()
    data = read_input(sys.argv[1])
    parse_time = time.perf_counter()

    p1 = part1(data, 50)
    p1_time = time.perf_counter()

    p2 = part2(data, 50)
    p2_time = time.perf_counter()

    total_time = (p2_time - start_time) * 1000
    p2_time = (p2_time - p1_time) * 1000
    p1_time = (p1_time - parse_time) * 1000
    parse_time = (parse_time - start_time) * 1000

    verbose = "\n"

    if verbose != "":
        print(f"{'parse':>15} ({parse_time:>10.4f}ms) ", end=verbose)

    print(f"{p1:>15} ({p1_time:>10.4f}ms) ", end=verbose)
    print(f"{p2:>15} ({p2_time:>10.4f}ms) ", end=verbose)
    print(f"{'total':>15} ({total_time:>10.4f}ms) ")


if __name__ == "__main__":
    main()

File: day01/solution.cpp
/* 2025 Advent of Code - Day 1
 *
 * Stephen Houser <stephenhouser@gmail.com>
 * All rights reserved.
 */

#include <getopt.h>	 // getopt

#include <algorithm>  // sort
#include <cassert>	  // assert macro
#include <chrono>	  // high resolution timer
#include <cstring>	  // strtok, strdup
#include <fstream>	  // ifstream (reading file)
#include <numeric>	  // max, reduce, etc.
#include <print>
#include <ranges>  // ranges and views
#include <string>  // strings
#include <vector>  // collectin

using namespace std;

/* Update with data type and result types */
using data_t = vector<int>;
using result_t = size_t;

/* for pretty printing durations */
using duration_t = chrono::duration<double, milli>;

/* Read the data file... */
const data_t read_data(const string& filename) {
	data_t data;

	std::ifstream ifs(filename);

	string line;
	while (getline(ifs, line)) {
		if (!line.empty()) {
			switch(line[0]) {
				case 'L':
					data.push_back(-stoi(line.substr(1)));
					break;
				case 'R':
					data.push_back(stoi(line.substr(1)));
					break;
				default:
					break;
			}
		}
	}

	return data;
}

/* Part 1 */
result_t part1(const data_t& data) {
	int position = 50;

	auto turn_dial = [&position](int n) {
		auto next = position;
		if (n < 0) {
			next += 100 * ((-n / 100) + 1);
		}
		position = (next + n) % 100;
		return position;
	};

	auto positions = data | views::transform(turn_dial);
	return (result_t)ranges::count(positions, 0);
}

result_t part2(const data_t& data) {
	int position = 50;

	/* Brute force solution */
	[[maybe_unused]]
	auto brute_force = [&position](int n) {
		auto zeros = 0;
		for (auto ticks = abs(n); ticks > 0; ticks--) {
			if (n < 0) {
				// + 100 to stay above 0
				position = (position - 1 + 100) % 100;
			} else {
				position = (position + 1) % 100;
			}

			zeros += position == 0 ? 1 : 0;
		}

		return zeros;
	};

	/* Original ugly math-y solution */
	[[maybe_unused]]
	auto ugly_crosses = [&position](int n) {
		auto next = position + n;

		if (next == 0) {
			position = next;
			return 1;
		}

		if (next < 0) {
			auto crossing = position == 0 ? -1 : 0;
			while (next < 0) {
				next += 100;
				crossing++;
			}

			crossing += next == 0 ? 1 : 0;

			position = next;
			return crossing;
		}

		auto crossing = (next / 100);
		position = next % 100;

		return crossing;
	};

	/* The math-y solution */
	auto elegant_crosses = [&position](int n) {
		if (n < 0) {
			position = 100 - position;
		}

		auto next = (position % 100) + abs(n);
		auto cross = next / 100;

		position = next % 100;

		if (n < 0) {
			position = 100 - position;
		}

		return cross;
	};

	auto crossings = data | views::transform(elegant_crosses);
	return std::accumulate(crossings.begin(), crossings.end(), 0u, std::plus());
}

int main(int argc, char* argv[]) {
	bool verbose = false;

	int c;
	while ((c = getopt(argc, argv, "v")) != -1) {
		switch (c) {
			case 'v':
				verbose = !verbose;
				break;
			default:
				std::print(stderr, "ERROR: Unknown option \"{}\"\n", c);
				exit(1);
		}
	}

	argc -= optind;
	argv += optind;

	const char* input_file = argv[0];
	if (argc != 1) {
		std::print(stderr, "ERROR: No input file specified\n");
		exit(2);
	}

	auto start_time = chrono::high_resolution_clock::now();

	auto data = read_data(input_file);

	auto parse_complete = chrono::high_resolution_clock::now();
	duration_t parse_time = parse_complete - start_time;
	if (verbose) {
		print("{:>15} ({:>10.4f}ms)\n", "parse", parse_time.count());
	}

	result_t p1_result = part1(data);

	auto p1_complete = chrono::high_resolution_clock::now();
	duration_t p1_time = p1_complete - parse_complete;
	print("{:>15} ({:>10.4f}ms){}", p1_result, p1_time.count(), verbose ? "\n" : "");

	result_t p2_result = part2(data);

	auto p2_complete = chrono::high_resolution_clock::now();
	duration_t p2_time = p2_complete - p1_complete;
	print("{:>15} ({:>10.4f}ms){}", p2_result, p2_time.count(), verbose ? "\n" : "");

	duration_t total_time = p2_complete - start_time;
	print("{:>15} ({:>10.4f}ms)\n", "total", total_time.count());
}
